### DISCUSSION TITLE
emptyDefaultForCollection

### DISCUSSION THREAD

**Discussion by cykl** [Randgalt/record-builder#90] [Ideas]
Immutables initializes collection attributes to the empty collection. I find this choice both practical and sensible. It avoids boilerplate and encourage to favor empty collections over null references (and a least distinguish the two concepts).

record-builder has emptyDefaultForOptional but I didn't find a way to initialize collection. Did i missed something? If not, could emptyDefaultForCollection option be added?

**Comment by Randgalt** [comment]
I guess it's not obvious but the option `useImmutableCollections` does this. If you build the project look for the generated `CollectionInterfaceRecordBuilder` to see what it generates.

**Reply by cykl** [reply]
So the problem is that is doesn't play nice with @Nonnull annotation.

```java
    @RecordBuilderFull
    record R(@Nonnull List<String> l) {
    }

    public static void main( String[ ] args ) {
        RemovemeRBuilder.builder( ).build( );
    }
```

throws

```
Exception in thread "main" java.lang.NullPointerException: l is required
	at java.base/java.util.Objects.requireNonNull(Objects.java:233)
	at com.greencomnetworks.RemovemeRBuilder.build(RemovemeRBuilder.java:78)
	at com.greencomnetworks.Removeme.main(Removeme.java:14)
```

Looking at generated source, it's now obvious that nullity check is performed just before invoking `__list(l)`.

```
    public Removeme.R build() {
        Objects.requireNonNull(l, "l is required");
        return new Removeme.R(__list(l));
    }
```

It should be an easy fix. Do you want me to create an issue or attempt a PR?

**Comment by Randgalt** [comment]
Yes, that would be appreciated!
